modal/image.py

```diff
@@ -1299,6 +1299,68 @@ def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
             gpu_config=parse_gpu_config(gpu),
         )
 
+    def uv_sync(
+        self,
+        lockfile: str = "./uv.lock",  # Path to lockfile
+        # Path to pyproject.toml. If not provided, then use file in the same directory as the lockfile
+        pyproject_toml: Optional[str] = None,
+        *,
+        force_build: bool = False,  # Ignore cached builds, similar to 'docker build --no-cache'
+        secrets: Sequence[_Secret] = [],
+        gpu: GPU_T = None,
+    ) -> "_Image":
+        """Install dependencies from a local `uv.lock` file using `uv sync`.
+
+        Note that only dependencies are installed. Please use `add_local_python_source` to include local python source
+        files. Please ensure that `modal` is a dependency by running `uv add modal`.
+        """
+
+        def build_dockerfile(version: ImageBuilderVersion) -> DockerfileSpec:
+            lockfile_ = os.path.expanduser(lockfile)
+            if pyproject_toml is None:
+                parent_dir = os.path.dirname(lockfile_)
+                pyproject_toml_ = os.path.join(parent_dir, "pyproject.toml")
+                if not os.path.exists(pyproject_toml_):
+                    msg = "pyproject.toml does not exist in the same directory as the uv.lock file"
+                    raise InvalidError(msg)
+            else:
+                pyproject_toml_ = os.path.expanduser(lockfile)
+
+            uv_root = "/.uv"
+            context_files = {"/.uv.lock": lockfile_, "/.pyproject.toml": pyproject_toml_}
+
+            uv_sync_args = [
+                f"--project={uv_root}",
+                "--frozen",  # Do not update `uv.lock`
+                "--no-install-workspace",  # Do not install the root project or any "uv workspace"
+                "--no-cache",  # Cache is not persisted, so we disable it
+                "--no-managed-python",  # Use the system python interpreter to create venv
+                "--compile-bytecode",
+            ]
```

Comment:
`uv sync` has like a million other options; how can we expose them for users who need to do something specific?

Reply:
Don't see any reason not to add it here. Are there any other commonly-set options that make sense to elevate into the method signature? We may also want to deconflict options we're hardcoding so users don't get confusing errors.

Reply:
Is it fair to say that `--group` is the main option there that would be useful for the "one `pyproject.toml` -> many image" setup that we'd like to support?

I'm also a little confused about the distinction between "extras" and "dependency groups"; the uv docs describe the latter as being primarily relevant to *development dependencies* but I'm not sure exactly how that would map onto a modal App, seems kind of aligned with the "non-published data science project" usecase they mention too.

Reply:
The "development dependency" is just a dependency group named "dev". It's the most common dependency group so they made it a first class citizen.

I can see one using `[project.optional-dependencies]` or "extras" to manage their different images. 

```toml
[project]
name = "my-app"
version = "0.1.0"
description = "An example of uv workspace"
readme = "README.md"
requires-python = ">=3.12"
dependencies = ["modal"]

[project.optional-dependencies]
extra1 = ["scipy>=1.15.3"]

[dependency-groups]
group1 = ["pydantic>=2.11.5"]

```

So one can either do `uv sync --group group1` or `uv sync --extra extra1`.

The biggest difference between them is that `[project.optional-dependencies]` is used to publish a package such that `my-lib[extra1]` works.

Reply:
You can do `uv add --group dev pytest` or `uv add --dev pytest`. They mean the same thing.

Reply:
Right I get the point about `--dev` aliasing, but that documentation about "development dependencies can be divided into multiple groups" makes me think they construe all `--groups` as "development" dependencies. Doesn't necessarily mean it's misaligned with use in a Modal App, just trying to make sure I understand how they're thinking about these features.